### PR TITLE
Mutation

### PR DESCRIPTION
**Pull request by oddoking**
add mutation random key and mutate prob

### CODE REVIEW

File: chromax/simulator.py [hunk 1]
```diff
@@ -150,6 +151,15 @@ def __init__(
...
                 f"One between {recombination_column} and {position_column} must be specified"
             )
+        if mutation is None:
+            self.mutation = 0.0
+        elif mutation > 0 and mutation < 1:
+            self.mutation = mutation
+        else:
+            raise ValueError(
+                f"mutation must be between 0 and 1, but got {mutation}"
+            )
```
Comment (younik) [inline chromax/simulator.py]: To simplify this, you can use as default 0 in init definition `mutation: float = 0f`

Then here, something like this:
```python
if mutation < 0 or mutation > 0:
   raise ValueError(
                 f"mutation must be between 0 and 1, but got {mutation}"
             )
self.mutation = mutation
```

File: chromax/simulator.py [hunk 2]
```diff
@@ -164,6 +174,8 @@ def set_seed(self, seed: int):
         :type seed: int
         """
         self.random_key = jax.random.PRNGKey(seed)
+        print(f'shape is {self.random_key.shape} {self.random_key[0]}')
+        return
```
Comment (younik) [inline chromax/simulator.py]: to remove

File: chromax/functional.py [hunk 3]
```diff
@@ -102,39 +120,51 @@ def double_haploid(
...
-    samples = jax.random.uniform(random_key, shape=recombination_vec.shape)
+    samples = jax.random.uniform(cross_random_key, shape=recombination_vec.shape)
     rec_sites = samples < recombination_vec
     crossover_mask = jax.lax.associative_scan(jnp.logical_xor, rec_sites)
 
     crossover_mask = crossover_mask.astype(jnp.int8)
     haploid = jnp.take_along_axis(individual, crossover_mask[:, None], axis=-1)
 
+    mutation_samples = jax.random.uniform(mutate_random_key, shape=haploid.shape)
+    mutation_sites = mutation_samples > mutate_probability
```
Comment (younik) [inline chromax/functional.py]: should be `mutation_samples < mutate_probability`, isn't it?
If `mutate_probability = 0`, `mutation_sites` should always be False

File: chromax/functional.py [hunk 4]
```diff
@@ -41,34 +46,44 @@ def cross(
...
-    random_keys = jax.random.split(random_key, num=len(parents) * 2 * parents.shape[3])
-    random_keys = random_keys.reshape(len(parents), 2, parents.shape[3], 2)
-    offsprings = _cross(parents, recombination_vec, random_keys)
+    cross_random_key = jax.random.split(cross_random_key, num=len(parents) * 2 * parents.shape[3])
+    cross_random_key = cross_random_key.reshape(len(parents), 2, parents.shape[3], 2)
+
+    mutate_split_key = jax.random.split(mutate_split_key, num=len(parents) * 2 * parents.shape[3])
+    mutate_split_key = mutate_split_key.reshape(len(parents), 2, parents.shape[3], 2)
+
+    offsprings = _cross(parents, recombination_vec, cross_random_key, mutate_split_key, mutate_probability)
```
Comment (younik) [inline chromax/functional.py]: I would avoid adding the other keys as argument, but I would generate them internally here.
So you can do:
```python
random_keys = jax.random.split(random_key, num=2 * len(parents) * 2 * parents.shape[3])
random_keys = random_keys.reshape(3, len(parents), 2, parents.shape[3], 2)
cross_random_key, mutate_random_key = random_keys
```

Also, use default value of 0 for mutate_probability, so previous code will continue working

File: chromax/functional.py [hunk 5]
```diff
@@ -41,34 +46,44 @@ def cross(
...
 
 
 def double_haploid(
     population: Population["n"],
     n_offspring: int,
     recombination_vec: Float[Array, N_MARKERS],
-    random_key: jax.random.PRNGKeyArray,
+    cross_random_key: jax.random.PRNGKeyArray,
+    mutate_random_key: jax.random.PRNGKeyArray,
+    mutate_probability: float,
```
Comment (younik) [inline chromax/functional.py]: same comment as `cross`

**Review by younik — CHANGES_REQUESTED** [review]
Many thanks for doing this!

I left some comments. I have time now, so if you need any help just let me know!

File: chromax/functional.py (hunk 4 above)
Comment (oddoking) [reply chromax/functional.py]: random_keys = random_keys.reshape(3, len(parents), 2, parents.shape[3], 2)

you mean 
random_keys = random_keys.reshape(2, len(parents), 2, parents.shape[3], 2)?

**Comment by oddoking** [comment]
have a look new change

**Comment by younik** [comment]
Merged, thank you! @oddoking

**Pull request merged**